### DISCUSSION TITLE
Shifting reference temperature to 0K to avoid negative energies

### DISCUSSION THREAD

**Discussion by capriatim** [mutationpp/Mutationpp#159] [General]
Some CFD codes may have problems when dealing with negative energies arising from setting the reference temperature at 298.15 K.
If you are using the RRHO database, the reference temperature can be simply shifted by comment out  '_- mp_part_sst[i]_' in the following function (in RrhoDB.cpp)
 
    _template <typename OP>
    void hF(double* const h, const OP& op) {
        LOOP(op(h[i], mp_hform[i] - mp_part_sst[i]))
    }_ 

This should be enough to shift the reference tempearure to 0K.
If you want to use the NASA-9 database, you should apply a shift in the computation of species enthalpy:
 h_i (T) | 0K =   h_i(T) | SST + h_shift
 where
 h_shift = hi(298.15)|SST - h(0K)|SST
 
 Please, @jbscoggi correct me if I am wrong.

**Comment by jbscoggi** [comment]
Hi @capriatim, actually I think there are a couple things wrong with your suggestion.  Indeed, removing the `-mp_part_sst[i]` as you suggested does correct the formation enthalpy for the RRHO model but it does not change the actual referenceTemperature().  To be consistent, I think this must also be changed to 0.  In the case of the NASA databases, simply shifting in the way you have described is also not really correct.  In general, the "reference species" which have formation enthalpies fixed at 0 (regardless of the reference conditions) should remain at zero.  I think what you have suggested will probably work in most cases as a bit of a "hack" but I'm afraid we should consider this issue more closely to make sure it is handled correctly in the future.

**Comment by capriatim** [comment]
Thank you @jbscoggi  for the comments.
To be conclusive for the RRHO part, one should also modify the initialization of the database (in RrhoDB.cpp)
_ThermoDB(298.15, 101325.0), m_ns(0), m_na(0), m_nm(0),_
For the NASA-9 polynomial we can discuss it next meeting to offer a more solid shift.
